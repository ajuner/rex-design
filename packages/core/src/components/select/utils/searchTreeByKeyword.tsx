import { isLeafNode } from 'ali-react-table';
import React from 'react';
import { TreeItem } from '../tree-view';

interface MatchResult {
  matched: boolean;
  startIndex: number;
  endIndex: number;
  input: string;
}

function match(item: TreeItem, keyword: string): MatchResult {
  if (typeof item.label === 'string') {
    const str = item.label.toLowerCase();
    const needle = keyword.toLocaleLowerCase();
    const index = str.indexOf(needle);

    return {
      matched: index > -1,
      startIndex: index,
      endIndex: index + needle.length,
      input: item.label,
    };
  }
  return { matched: false, startIndex: 0, endIndex: 0, input: (item.label as unknown) as string };
}

function decorate(matchResult: MatchResult) {
  const { input, startIndex, endIndex, matched } = matchResult;

  if (!matched) {
    return input;
  }

  return (
    <span>
      {input.slice(0, startIndex)}
      <span style={{ margin: 1, fontWeight: 'bold', color: '#333', background: '#fff8bd' }}>
        {input.slice(startIndex, endIndex)}
      </span>
      {input.slice(endIndex)}
    </span>
  );
}

const dfsResultOfLeafNode: DfsResult = {
  anyItemMatched: false,
  anyDescendentMatched: false,
  list: [],
};

interface DfsResult {
  list: TreeItem[];
  anyItemMatched: boolean;
  anyDescendentMatched: boolean;
}

/**
 * 根据关键字 `keyword` 在树中进行搜索，返回结果为一个对象，表示本次搜索的结果。
 *
 * * `skipped` 搜索是否被跳过了（keyword 为空时搜索将被跳过）
 * * `filteredDataSource` 过滤之后的数据源
 * * `preferredExpandedKeys` 根据关键字计算得到的理想的 expandedKeys
 *
 * */
export function searchTreeByKeyword(
  untrimmedKeyword: string,
  dataSource: TreeItem[],
): {
  skipped: boolean;
  filteredDataSource: TreeItem[];
  preferredExpandedKeys: string[];
} {
  const keyword = untrimmedKeyword.trim();
  if (keyword === '') {
    return {
      filteredDataSource: dataSource,
      preferredExpandedKeys: [],
      skipped: true,
    };
  }

  const ctx = {
    expandedKeys: new Set<string>(),
  };
  const dfsResult = dfs(ctx, dataSource, false);

  return {
    skipped: false,
    filteredDataSource: dfsResult.list,
    preferredExpandedKeys: Array.from(ctx.expandedKeys),
  };

  function dfs(ctx: { expandedKeys: Set<string> }, items: TreeItem[], anyAncestorMatched: boolean): DfsResult {
    let anyDescendentMatched = false;
    let anyItemMatched = false;
    const result: TreeItem[] = [];

    for (const item of items) {
      const selfMatchResult = match(item, keyword);

      const subDfsResult: DfsResult = isLeafNode(item)
        ? dfsResultOfLeafNode
        : dfs(ctx, item.children, selfMatchResult.matched || anyAncestorMatched);

      const needIncludeInResult = anyAncestorMatched || selfMatchResult.matched || subDfsResult.anyDescendentMatched;

      if (needIncludeInResult) {
        result.push({
          ...item,
          label: decorate(selfMatchResult),
          children: subDfsResult.list,
        });
      }

      if (subDfsResult.anyDescendentMatched) {
        ctx.expandedKeys.add(item.key);
      }

      anyItemMatched ||= selfMatchResult.matched;
      anyDescendentMatched ||= selfMatchResult.matched || subDfsResult.anyDescendentMatched;
    }

    return { list: result, anyItemMatched, anyDescendentMatched };
  }
}
