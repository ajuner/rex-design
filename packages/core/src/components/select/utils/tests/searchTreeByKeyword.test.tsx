import { searchTreeByKeyword } from '../searchTreeByKeyword';
import { bigTreeDataSource } from './test-data';

test('basic search', () => {
  const result = searchTreeByKeyword('Option 1', [
    { key: '1', label: 'Option 1' },
    { key: '2', label: 'Option 2' },
    { key: '3', label: 'Option 3' },
  ]);

  expect(result.filteredDataSource.length).toBe(1);
  expect(result.filteredDataSource[0].key).toBe('1');
  expect(result.skipped).toBe(false);
  expect(result.preferredExpandedKeys).toEqual([]);
});

test('skip search', () => {
  const result = searchTreeByKeyword('', [
    { key: '1', label: 'Option 1' },
    { key: '2', label: 'Option 2' },
    { key: '3', label: 'Option 3' },
  ]);

  expect(result.skipped).toBe(true);
});

test('search big tree', () => {
  expect(searchTreeByKeyword('小灰', bigTreeDataSource)).toMatchSnapshot();
});
