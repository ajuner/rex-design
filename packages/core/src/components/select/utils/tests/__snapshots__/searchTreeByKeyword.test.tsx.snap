// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search big tree 1`] = `
Object {
  "filteredDataSource": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "小灰灰",
              "label": <span>
                
                <span
                  style={
                    Object {
                      "background": "#fff8bd",
                      "color": "#333",
                      "fontWeight": "bold",
                      "margin": 1,
                    }
                  }
                >
                  小灰
                </span>
                灰
              </span>,
            },
          ],
          "key": "foo-4",
          "label": "其他小朋友",
        },
      ],
      "key": "zoo",
      "label": "阿里动物园",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "key": "bar-4-小灰灰",
              "label": <span>
                
                <span
                  style={
                    Object {
                      "background": "#fff8bd",
                      "color": "#333",
                      "fontWeight": "bold",
                      "margin": 1,
                    }
                  }
                >
                  小灰
                </span>
                灰
              </span>,
            },
          ],
          "key": "bar-4",
          "label": "小朋友",
        },
      ],
      "key": "2号动物园",
      "label": "2号动物园",
    },
  ],
  "preferredExpandedKeys": Array [
    "foo-4",
    "zoo",
    "bar-4",
    "2号动物园",
  ],
  "skipped": false,
}
`;
